linkerd/service-profiles/src/split.rs

```diff
@@ -93,116 +132,84 @@ where
-        }
+        match self.inner {
+            Inner::Default(ref mut svc) => svc.poll_ready(cx).map_err(Into::into),
+            Inner::Split {
+                ref mut rx,
+                ref mut services,
+                ref mut addrs,
+                ref mut distribution,
+                ref mut new_service,
+                ref target,
+                ..
+            } => {
+                let mut update = None;
+                while let Poll::Ready(Some(up)) = rx.poll_recv_ref(cx) {
+                    update = Some(up.clone());
+                }
 
-        // If, somehow, the watch hasn't been notified at least once, build the
-        // default target. This shouldn't actually be exercised, though.
-        if self.inner.is_none() {
-            self.update_inner(Vec::new());
-        }
-        debug_assert_ne!(self.services.len(), 0);
+                // Every time the profile updates, rebuild the distribution, reusing
+                // services that existed in the prior state.
+                if let Some(Profile { mut targets, .. }) = update {
+                    if targets.len() == 0 {
+                        targets.push(Target {
+                            addr: target.into(),
+                            weight: 1,
+                        })
+                    }
+                    debug!(?targets, "Updating");
+
+                    let mut prior_addrs =
+                        std::mem::replace(addrs, IndexSet::with_capacity(targets.len()));
+                    let mut weights = Vec::with_capacity(targets.len());
+
+                    // Create an updated distribution and set of services.
+                    for Target { weight, addr } in targets.into_iter() {
+                        // Reuse the prior services whenever possible.
+                        if !prior_addrs.remove(&addr) {
+                            debug!(%addr, "Creating target");
+                            let svc = new_service.new_service((Some(addr.clone()), target.clone()));
+                            services.push(addr.clone(), svc);
+                        } else {
+                            trace!(%addr, "Target already exists");
+                        }
+                        addrs.insert(addr);
+                        weights.push(weight);
+                    }
+
+                    *distribution = WeightedIndex::new(weights).unwrap();
+
+                    for addr in prior_addrs.into_iter() {
+                        services.evict(&addr);
+                    }
+                }
 
-        // Wait for all target services to be ready. If any services fail, then
-        // the whole service fails.
-        Poll::Ready(ready!(self.services.poll_pending(cx)).map_err(Into::into))
+                // Wait for all target services to be ready. If any services fail, then
+                // the whole service fails.
+                Poll::Ready(ready!(services.poll_pending(cx)).map_err(Into::into))
```

Comment:
this logic used to all live in `update_inner` and was just moved here as a result of `inner` becoming an enum, right? the diff didn't make that super obvious so i wanted to double check

Reply:
There's no functional change, but the block has been cleaned up slightly (mostly pushing a default target on to targets when there are none) to reduce duplication.